### ISSUE TITLE
flux-config problems for installed flux

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#490]
After `make install` was fixed up in #488 I found a few issues with the installed version of `flux` -- all of which *may* be related to the installed flux-config handling.

I actually don't remember how this is supposed to work, but thought I'd post here first to discuss while I'm investigating further.

BTW, I'm testing against an installed version of flux-core by first configuring with a writable install prefix
```
$ ./configure --prefix=/tmp/grondo/flux
```

Then building and installing
```
$ make && make install
```

Then pointing the sharness test path to the installed `flux(1)` instead of the in-tree version (the default)
```
$ cd t
$ FLUX_TEST_INSTALLED_PATH=/tmp/grondo/flux/bin make check
```

Note that part of the problem could be related to usage of `FLUX_TEST_INSTALLED_PATH`, which itself is not well tested. 

The first failure is the basic `flux-config` test:
```
not ok 6 - flux-config works
#	
#		flux config get general.broker_path
#	
```

running directly:
```
$  /tmp/grondo/flux/bin/flux config get general.broker_path
flux-config: general.broker_path: No such file or directory
```

This is because there is no config:
```
$ /tmp/grondo/flux/bin/flux config dump
$
```

`FLUX_CONFIG_DIR` seems to be set to `~/.flux` -- is that right?
```
$ /tmp/grondo/flux/bin/flux env | grep DIR
FLUX_CONF_DIRECTORY=/g/g0/grondo/.flux
FLUX_SEC_DIRECTORY=/g/g0/grondo/.flux
```

Seems like there should be a `flux/config` under `$sysconfdir`, but `etc/flux/flux.conf.in` is purposely not generated for install, so probably I don't understand how this is supposed to work?

To add to the confusion, Lua and Python bindings paths seem to be correct for the installed flux, because they are apparently generated from automake substitutions (and overridden perhaps by config).

Is the fix here that all items in the `config` should also have automake substituted or `#define`'d defaults?

Other test failures possibly related to `flux config` is that recursive flux doesn't work under my installed flux:
```
(flux-121076) grondo@ipa15:/tmp/grondo/flux$ flux wreckrun -n2 flux broker
[1449687573.028322] job.info[0]: got request job.create
[1449687573.028529] job.info[0]: Setting job 1 to reserved
[1449687573.043725] lwj.1.info[0]: lwj 1: node0: nprocs=1, nnodes=2, cmdline=[ "flux", "broker" ]
[1449687573.048528] lwj.1.info[1]: lwj 1: node1: nprocs=1, nnodes=2, cmdline=[ "flux", "broker" ]
broker: could not load config from KVS: No such file or directory
broker: could not load config from KVS: No such file or directory
[1449687573.063985] lwj.1.info[1]: job complete. exiting...
[1449687573.065902] lwj.1.info[0]: job complete. exiting...
wreckrun: tasks [0-1]: exited with exit code 1
```

**Comment by grondo** [comment]
Looking into this a little further the problems are

 * Installed `flux` does not have a config file by default. **Some** config items are set internally, but without an actual config file present `flux-config` and presumably the kvs-upload code ignores these values.

 * When there is no config *file*, flux config items are not pushed into the kvs of an instance, which seems to violate an invariant assumed by code, causes `flux-config` errors, and of course `flux-broker` doesn't work as shown above
 ```
$ bin/flux start -o -L/dev/null flux config dump
flux-config: could not load config from KVS: No such file or directory
flux-start: 0 (pid 148011) exited with rc=1
 ```

In the broker, it isn't a fatal error if `FLUX_CONF_USEFILE` is set but the file doesn't exist, but the same courtesy is not extended to the KVS config:
```C
    if (getenv ("FLUX_CONF_USEFILE")) {
        if (ctx.verbose)
            msg ("Loading config from %s", confdir);
        if (flux_conf_load (ctx.cf) < 0 && errno != ESRCH)
            err_exit ("%s", confdir);
    } else if (getenv ("FLUX_URI")) {
        flux_t h;
        if (ctx.verbose)
            msg ("Loading config from KVS");
        if (!(h = flux_open (NULL, 0)))
            err_exit ("flux_open");
        if (kvs_conf_load (h, ctx.cf) < 0)
            err_exit ("could not load config from KVS");
        flux_close (h);
```

If the KVS config should *always* exist, then the broker should do that even if no conffile is present, otherwise, perhaps it is ok to allow `kvs_conf_load()` to fail similar to `flux_conf_load()`. That might be a good short term fix while we redesign the whole conf structure  (I would actually prefer a single `flux_conf_load()` that loads in a predictable manner, with perhaps `internal conf < kvs conf < file conf < environment conf < cmdline` (where `<` means overridden by...)?

possible short term workaround? 
```diff
diff --git a/src/broker/broker.c b/src/broker/broker.c
index e677996..cbbd143 100644
--- a/src/broker/broker.c
+++ b/src/broker/broker.c
@@ -433,7 +433,7 @@ int main (int argc, char *argv[])
             msg ("Loading config from KVS");
         if (!(h = flux_open (NULL, 0)))
             err_exit ("flux_open");
-        if (kvs_conf_load (h, ctx.cf) < 0)
+        if (kvs_conf_load (h, ctx.cf) < 0 && errno != ENOENT)
             err_exit ("could not load config from KVS");
         flux_close (h);
     }
```

**Comment by garlick** [comment]
See also #452

**Comment by grondo** [comment]
Thanks! Let's track quick fix in this issue, and I'll move any comments related to config overhaul or rewrite into #452.

**Comment by grondo** [comment]
Ok the immediate problems described here were fixed in PR #494. Closing this issue and moving config rework discussion over to #452.
